### ISSUE TITLE
PHP files do not work in app/Resources/contao/languages

### ISSUE THREAD

**Issue by fritzmg** [contao/core-bundle#740]
You are supposed to be able to provide your own translation files within `app/Resources/contao/languages`. In theory, you should be able to provide either XLF or PHP files there, for example
```
app/Resources/contao/languages/de/default.xlf
```
with
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xliff version="1.1">
  <file datatype="php" original="src/Resources/contao/languages/en/default.php" source-language="en">
    <body>
      <trans-unit id="MSC.more">
        <target>mehr »</target>
      </trans-unit>
    </body>
  </file>
</xliff>
```
or
```
app/Resources/contao/languages/de/default.php
```
with
```php
<?php

$GLOBALS['TL_LANG']['MSC']['more'] = 'mehr »';
```
However, the latter does not seem to work.

_Note:_ I deleted the `dev` and `prod` cache after each change of course.

See also: https://community.contao.org/de/showthread.php?66223-Weiterlesen-Link-unter-Contao-4-umbenennen

**Comment by leofeyer** [comment]
@aschempp /cc

**Comment by aschempp** [comment]
I cannot confirm this atm. I'm actively using PHP language files in `app/Resources` and it works just fine…

**Comment by discordier** [comment]
So am I, I have not a single xlf in my extension code but only xlf.
Therefore we need someone to confirm and pinpoint the problem.

**Comment by aschempp** [comment]
You know this is not about extensions, right?

**Comment by mvscheidt** [comment]
The files are loaded indeed. I purposely caused a Synthax error in my file (default.php) and it threw an error there after clearing the cache. So this is most likely not an issue.

However the loading order is important here.
Both the PHP Files AND the Xliff Loader set the $GLOBALS Array currently. The latter because the constructors second param is true: https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/System.php#L376

The "PHP Loader": https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/System.php#L390-L402

The xlfLoader Loader however happens to run after the PHP Files are loaded (you probably see where this is going).

Xliff Loader: https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/System.php#L404-L416

If you follow the different calls coming from `load` https://github.com/contao/core-bundle/blob/master/src/Config/Loader/XliffFileLoader.php#L53-L56 you will end up at `\Contao\CoreBundle\Config\Loader\XliffFileLoader::getPhpFromFileNode` which then calls `\Contao\CoreBundle\Config\Loader\XliffFileLoader::addGlobal` https://github.com/contao/core-bundle/blob/master/src/Config/Loader/XliffFileLoader.php#L230-L242.

This method then  "adds" the loaded xliff translation to the globals array. Except it doesn't check if an entry exists and thus ends up overwriting existing $GLOBALS keys.

**tl;dr**

Basically the xliff files of the core get loaded after the custom .php files, causing custom changes to Core translations to be overridden.

Though I am still in the middle of sorting this out. I did have a similar problem however when I was developing an extension and trying to overwrite a core translation in my extensions php file (didn't work until I created xlf files)

**Comment by discordier** [comment]
> You know this is not about extensions, right?

It should not matter where the language file is loaded from, should it? Be it a vendor language resource dir or app resources is the same code in the end.

> I did have a similar problem however when I was developing an extension and trying to overwrite a core translation in my extensions php file (didn't work until I created xlf files)

Have you ensured your bundle is loaded after the core? Maybe that is the reason.

**Comment by mvscheidt** [comment]
Yes, it is loaded after the Contao Bundles in AppKernel.php.

Though that was just added information (where I noticed it first, mainly) - I currently face the same issue as @fritzmg with `default.php` inside `src/sites/fvf/app/Resources/contao/languages/de` - trying to overwrite some core translations.

**Comment by DanielSchwiperich** [comment]
I think it's pretty clear in the implementation linked by @mvscheidt It first loads all php files and after that all the xlf files. (with two separate locate calls on the conto resource locator)


And since the core bundle comes with xlf files the only chance to overwrite a core translation is by another xlf files (this works because of the dir load order given to the locator)
Or to use the deprecated langconfig.php

**Comment by mvscheidt** [comment]
Example - I'm using the Registration Module of Contao - the Customer wants the "Bestätigung" (of the Password Widget) to be changed to a different wording, so I did this:

app/Resources/contao/languages/de/default.php
```
<?php
$GLOBALS['TL_LANG']['MSC']['confirmation'] = ['Passwort Wiederholen'];
```

This does not work, even after clearing the Cache via bin/console - the Registration Module keeps showing "Bestätigen"

Adding

app/Resources/contao/languages/de/default.xlf
```
<?xml version="1.0" encoding="UTF-8"?>
<xliff xmlns="urn:oasis:names:tc:xliff:document:1.2" version="1.2">
  <file datatype="php" date="2017-03-22T12:28:38+01:00" original="default" source-language="en" target-language="de">
    <body>
      <trans-unit id="MSC.confirmation">
        <source>Confirmation</source>
        <target>Passwort Wiederholen</target>
      </trans-unit>
    </body>
  </file>
</xliff>
```

And then it works. Removing the php file and clearing the cache makes no difference at all. As long as the xlf file is there it works properly

**Comment by fritzmg** [comment]
Btw. just for the record: what does not work is overriding existing XLF translations with `.php` files. Generating new translations does work.

**Comment by Samson1964** [comment]
Ich will in C 4.4.7 die modules.xlf einer C3-Erweiterung überschreiben. Das Funktioniert aber nicht. Hat die C3-Erweiterung dagegen eine modules.php als Sprachdatei, funktioniert es einwandfrei. Die Sprachdateien mit dem Prefix tl_ scheinen aber alle zu klappen, auch die mit xlf als Suffix. Die dafult.xlf habe ich nicht getestet. 
Siehe auch: [Forum](https://community.contao.org/de/showthread.php?68593-Sprachvariablen-%C3%A4ndern)

**Comment by leofeyer** [comment]
So how can this be reproduced?

**Comment by Toflar** [comment]
Again: You cannot override `modules.xlf` with `modules.php`.

**Comment by fritzmg** [comment]
So this is a known limitation? Why does it work for other files, but not for modules.xlf?

**Comment by Toflar** [comment]
Go check the `System::loadLanguageFile()`. It currently uses the `contao.resource_locator` service that is an instance of `Symfony\Component\Config\FileLocator`. It does not support finding files with multiple file extensions. But this would be needed so we can guarantee the loading order.

**Comment by fritzmg** [comment]
@Toflar yes, but what is special about modules.xlf vs. default.xlf or tl_content.xlf etc?

**Comment by Toflar** [comment]
Nothing is special? They‘re just loaded after all php files.

**Comment by fritzmg** [comment]
No I meant - why can you override language strings in any other `.php` file, but not language strings, that are defined in the `module.xlf`? That is the bug we are talking about in this issue.

**Comment by Toflar** [comment]
As I said, have a look at `System::loadLanguageFile()`. It works because `.php` files are loaded first, then `.xlf`.

**Comment by fritzmg** [comment]
I don't get it ;). If PHP files are loaded first and then XLF files, wouldn't that mean that you cannot override _any_ languages strings with PHP files, if they were first defined in XLF files?

**Comment by Toflar** [comment]
That's correct.

**Comment by fritzmg** [comment]
But that is not the case.

**Comment by Toflar** [comment]
Just check the code please, I cannot tell you more than the code.

**Comment by fritzmg** [comment]
I know what the code says, but it's not actually working this way.

## Example 1: modules.php in app/Resources

_app/Resources/contao/languages/de/modules.php_
```php
<?php

$GLOBALS['TL_LANG']['MOD']['content'] = 'Lorem ipsum';
```
![screenshot-2017-12-19 benutzerprofil - contao open source cms](https://user-images.githubusercontent.com/4970961/34160353-e8b1e45e-e4cc-11e7-90ae-ec8754ca8e9c.png)

## Example 2: modules.php in system/modules/my_extension

_system/modules/my_extension/languages/de/modules.php_
```php
<?php

$GLOBALS['TL_LANG']['MOD']['content'] = 'Dolor sit';
```
![screenshot-2017-12-19 startseite - contao open source cms](https://user-images.githubusercontent.com/4970961/34160536-9d03b8ba-e4cd-11e7-86f5-a8bcd2d35c12.png)

This means you __can__ generally override language strings in `.php` files even if the original language string was defined in an `.xlf` file.

**Comment by Toflar** [comment]
That probably only works when the cache is built (which is also a php file).

**Comment by fritzmg** [comment]
Of course, I did a `cache:clear` after these changes.

**Comment by Toflar** [comment]
Yeah, so it builds the internal cache (xlf's are all converted and combined to one php file) which means you can then override stuff that was originally in an `xlf` file.

**Comment by fritzmg** [comment]
So this explains the original confusion about this issue. When I first tested it, the internal cache of Contao probably was not built.

**Comment by MacKP** [comment]
WoW this was realy confusing for me... I had to change some things in my own AppBundle like:
```
$GLOBALS['TL_LANG']['MSC']['male']             = 'Herr';
$GLOBALS['TL_LANG']['MSC']['female']           = 'Frau';
$GLOBALS['TL_LANG']['tl_member']['gender'] = ['Anrede', 'Bitte wählen Sie Ihre Anrede.'];
```

The first two lines worked with app_dev.php but the last line not. The last line was only shown without app_dev.php that tricked me realy hard ;-)
Would be nice, if there would be an solution for this.
regards

**Comment by leofeyer** [comment]
@MacKP A solution for what exactly?

**Comment by MacKP** [comment]
That there is shown something else when you use app_dev.php and withoug it.

regards

**Comment by aschempp** [comment]
What file did you add in your AppBundle? You can usually not mix `MSC` and `tl_member` in the same file.

**Comment by fritzmg** [comment]
@aschempp Reproduction steps:

1. Create `app/Resources/contao/languages/de/tl_member.php`.
2. Insert `$GLOBALS['TL_LANG']['tl_member']['gender'] = ['Anrede', 'Bitte wählen Sie Ihre Anrede.'];`.
3. Access the Contao installation via `app_dev.php`.

The changes will not be visible. If you clear the production cache and access the Contao installation via `app.php`, the changes will be visible.

_In short:_ changing language strings in `.php` files does not work in the `dev` environment, because the `.xlf` files are loaded after the `.php` files. It works in the `prod` environment due to the language cache.

**Comment by aschempp** [comment]
What file did you add in your AppBundle? You can usually not mix `MSC` and `tl_member` in the same file.

**Comment by aschempp** [comment]
This is a known problem, you should use the same format to override a language file. I agree this is cumbersome though…

**Comment by fritzmg** [comment]
Yes it's a known limitation currently, but I also agree that it is cumbersome. I never go through the effort of creating an XLF file just to replace one or two language strings - and just so that I can see the change via `dev` as well ;)

**Comment by Toflar** [comment]
It is due to the file loader not being able to load different formats at the same time. But maybe that changed in Symfony lately because they also support wildcard globbing now. I also think we should try to fix that if possible. Any takers?

**Comment by DanielSchwiperich** [comment]
@mvscheidt Let's try to work on this next week

**Comment by mvscheidt** [comment]
Just documenting the current behavior I encountered in a project based on contao `4.4.18`:

I tried to overwrite the `$GLOBALS['TL_LANG']['MSC']['more']` Translation for `cs` via php file in `app/Resources/contao/languages/cs/default.php`

**Empty Cache Folder**

 * Using `app.php` - translation was not applied
   * I did not expect this - I assumed when calling the page the language cache would be created. But it's not so you need to explicitly warm the cache.
 * Using `app_dev.php` - translation was not applied 
   * current behavior according to this ticket

**Warming Cache**

 (makes no difference if you call the page in Browser first and warm afterwards or vice versa, the below behaviour is the same - this applies for both environments)

 * `bin/console cache:warmup --env=prod` using `app.php` - translation is applied
   * Works as expected according to the ticket
 * `bin/console cache:warmup --env=dev` using `app_dev.php` - translation is applied
   * I did not expect this, from what I understood in this ticket this should not work currently?

**Comment by NinaG** [comment]
Dazu ergänzend von meiner Seite - wenn ich die Stelle aus dem Beispiel in diesem Thread via default.php statt default.xlf ändere, konnte ich die Änderung durch klassisches Leeren des Prod-Caches nie sehen. Habe ich es stattdessen per default.xlf gemacht, war die Änderung da.
https://community.contao.org/de/showthread.php?71992-Phrasen-in-Contao-4-ändern-(früher-langconfig-php)&p=482581

**Comment by mvscheidt** [comment]
@NinaG 

Könntest du so nett sein und ausprobieren ob deine `.php` Übersetzung geladen wird wenn du nach dem Cache-clear den Befehl `vendor/bin/contao-console cache:warmup -e prod` ausführst?

Das sollte meiner Meinung nach klappen und würde sich dann mit meinem vorherigen Kommentar decken, das dieses Problem nur auftritt, wenn man den Cache NICHT vorwärmt.

Ich sehe in einer verlinkten Beschreibung das du zwar beschreibst den Cache zu leeren, aber nicht anmerkst diesen auch wieder aufbauen zu lassen.

Das könntest du dann zwischenzeitlich für altere Contao Versionen als "möglichen" Workaround für das Problem hinterlegen - wobei ich nicht weiß bis zu welcher Version das klappen würde.

Davon abgesehen sollte das Cache Warming in Produktivumgebungen nach einem Deploy ohnehin gemacht werden.

Machst du das nicht, so existiert unter anderem der Contaos Cache (oder zumindest ein Teil davon) nicht.

Das ist daran ersichtlich das der `var/cache/prod/contao` Ordner fehlt - welcher anscheinend nur vom Cache Warmup Befehl angelegt wird.

**Comment by leofeyer** [comment]
Finally fixed in contao/contao#110 thanks to @mvscheidt.
